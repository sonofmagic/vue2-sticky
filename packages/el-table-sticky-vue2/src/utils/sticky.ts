import type { ObjectDirective, VNode } from 'vue'
import { checkElTable, convertToPx } from './index'
import Scroller from './scroller'

export interface StickyOptions {
  offsetTop?: number
  offsetBottom?: number
}

const defaultScrollbarHeight = 10
/**
 * @class Sticky
 * @classdesc sticky header or footer for el-table
 * @abstract
 */
export default class Sticky {
  /**
   * from new.target.name
   * @type {string}
   */
  target: string

  offsetTop!: string
  offsetBottom!: string

  /**
   * Constructor for StickyHeader and StickyFooter
   * @param {object} options options from Vue.use
   * @param {number | string} [options.offsetTop] the top offset of the table header
   * @param {number | string} [options.offsetBottom] the bottom offset of the table footer
   */
  constructor({ offsetTop = 0, offsetBottom = 0 }) {
    this.target = new.target.name
    if (this.target === 'StickyHeader') {
      this.offsetTop = convertToPx(offsetTop)
      this.offsetBottom = convertToPx(offsetBottom) // for horizontal scrollbar
    }
    if (this.target === 'StickyFooter') {
      this.offsetBottom = convertToPx(offsetBottom)
    }
  }

  resetSticky(tableCell: HTMLElement[] = []) {
    // reset data-sticky attribute and style
    tableCell.forEach((th) => {
      th.removeAttribute('data-sticky')
      th.removeAttribute('style')
    })
  }

  /**
   * Stack sticky for left fixed columns
   * @param {Array} tableCell table header or footer cells
   */
  stackLeftColumns(tableCell: HTMLElement[] = []) {
    let stickyLeft = 0
    for (let i = 0; i < tableCell.length; i++) {
      const th = tableCell[i]
      if (th.classList.contains('is-hidden')) {
        th.dataset.sticky = 'left'
        th.style.left = `${stickyLeft}px`
        stickyLeft += th.offsetWidth
        // set data-sticky-left-last attribute for last left fixed column
        if (!th.nextElementSibling?.classList.contains('is-hidden')) {
          th.dataset.sticky = 'end'
          break
        }
      }
      else if (i === 0) {
        // no left fixed columns
        break
      }
    }
  }

  /**
   * Stack sticky for right fixed columns
   * @param {Array} tableCell table header or footer cells
   */
  stackRightColumns(tableCell: HTMLElement[] = []) {
    let stickyRight = 0
    for (let i = tableCell.length - 1; i >= 0; i--) {
      const th = tableCell[i]
      if (th.classList.contains('is-hidden') && !th.dataset.sticky) {
        th.dataset.sticky = 'right'
        th.style.right = `${stickyRight}px`
        stickyRight += th.offsetWidth
        // set data-sticky-right-first attribute for first right fixed column
        if (!th.previousElementSibling?.classList.contains('is-hidden')) {
          th.dataset.sticky = 'start'
          break
        }
      }
    }
  }

  getStickyWrapperCells(el: Element, binding: { value: StickyOptions }): HTMLElement[] | undefined {
    const { value } = binding
    let selector, styleProperty, offsetProperty

    if (this.target === 'StickyHeader') {
      selector = '.el-table__header'
      styleProperty = 'top'
      offsetProperty = 'offsetTop'
    }

    if (this.target === 'StickyFooter') {
      selector = '.el-table__footer'
      styleProperty = 'bottom'
      offsetProperty = 'offsetBottom'
    }

    const tableStickyWrapper = el.querySelector(`${selector}-wrapper`)
    if (tableStickyWrapper) {
      tableStickyWrapper.style[styleProperty] = value?.[offsetProperty] !== void 0
        ? convertToPx(value[offsetProperty])
        : this[offsetProperty]
      // element-ui 2.15.0 是没有这个 .el-table__cell 的
      // 但是 .el-table__cell 在 2.15.14 版本是有的
      const selectorAll = `${selector} ${this.target === 'StickyHeader' ? 'th' : 'td'}`
      return Array.from(tableStickyWrapper.querySelectorAll(selectorAll))
    }
  }

  /**
   * Init scroller for el-table v-sticky-header or v-sticky-footer
   */
  async initScroller(el: Element, binding: { value: StickyOptions }, vnode: VNode) {
    const { value } = binding
    if (vnode.componentInstance && vnode.componentInstance.layout) {
      // windows 下必须设置这个配置
      vnode.componentInstance.layout.gutterWidth = value?.scrollbarHeight ?? defaultScrollbarHeight
    }

    const scrollerOffsetBottom = value?.offsetBottom !== void 0 ? convertToPx(value.offsetBottom) : this.offsetBottom
    if (this.target === 'StickyFooter' && el.scroller) {
      // wait for el-table render
      await vnode.componentInstance.$nextTick()
      el.scroller.scrollbar?.destroy()
      el.scroller = null
    }
    el.scroller = new Scroller(el, binding, vnode, scrollerOffsetBottom)
  }

  async stackStickyColumns(el: Element, binding: { value: StickyOptions }, vnode: VNode, reset = false) {
    // wait for el-table render
    await vnode.componentInstance?.$nextTick()

    const tableCell = this.getStickyWrapperCells(el, binding)

    reset && this.resetSticky(tableCell)
    this.stackLeftColumns(tableCell)
    this.stackRightColumns(tableCell)
  }

  /**
   * Init directive config for Vue
   * @returns {object} directive config
   */
  init(): ObjectDirective {
    return {
      inserted: (el, binding, vnode) => {
        checkElTable(binding, vnode)
        // set data-sticky-* attribute for el-table
        el.dataset[this.target.replace(/^\S/, s => s.toLowerCase())] = ''
        this.initScroller(el, binding, vnode)
        this.stackStickyColumns(el, binding, vnode)
      },
      update: (el, binding, vnode) => {
        this.stackStickyColumns(el, binding, vnode, true)
      },
      unbind: (el) => {
        if (el.scroller) {
          el.scroller.scrollbar?.destroy()
          el.scroller = null
        }
      },
    }
  }
}
