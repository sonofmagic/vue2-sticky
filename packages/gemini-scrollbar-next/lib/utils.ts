let _scrollbarWidth: number

function getScrollbarWidth() {
  if (_scrollbarWidth !== undefined) {
    return _scrollbarWidth
  }
  const e = document.createElement('div')
  // 但现代浏览器有批处理优化，这种差距几乎察觉不到。
  e.style.position = 'absolute'
  e.style.top = '-9999px'
  e.style.width = '100px'
  e.style.height = '100px'
  e.style.overflow = 'scroll'
  document.body.appendChild(e)
  _scrollbarWidth = e.offsetWidth - e.clientWidth
  e.remove()
  return _scrollbarWidth
}

function addClass(el: HTMLElement, classNames: string[]) {
  el.classList.add(...classNames)
}

function removeClass(el: HTMLElement, classNames: string[]) {
  el.classList.remove(...classNames)
}

export {
  addClass,
  getScrollbarWidth,
  // isIE,
  removeClass,
}
