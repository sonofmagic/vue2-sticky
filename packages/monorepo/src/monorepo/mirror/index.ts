export * from './sources'
export * from './utils'
