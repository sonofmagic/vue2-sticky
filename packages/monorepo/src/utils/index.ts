export * from './md5'
export * from './regexp'
