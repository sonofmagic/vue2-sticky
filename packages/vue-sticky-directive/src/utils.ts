import { defu } from 'defu'

export {
  defu,
}
