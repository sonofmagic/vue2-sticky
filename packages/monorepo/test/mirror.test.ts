import { setMirror } from '@/monorepo/mirror/utils'

describe('mirror', () => {
  it('setMirror', () => {
    const settingsJson = {}
    setMirror(settingsJson)
    expect(settingsJson).toMatchSnapshot()
  })
})
