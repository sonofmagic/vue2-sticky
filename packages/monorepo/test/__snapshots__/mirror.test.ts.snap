// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mirror > setMirror 1`] = `
{
  "terminal.integrated.env.linux": {
    "CHROMEDRIVER_CDNURL": "https://cdn.npmmirror.com/binaries/chromedriver",
    "COREPACK_NPM_REGISTRY": "https://registry.npmmirror.com",
    "CYPRESS_DOWNLOAD_PATH_TEMPLATE": "https://cdn.npmmirror.com/binaries/cypress/\${version}/\${platform}-\${arch}/cypress.zip",
    "EDGEDRIVER_CDNURL": "https://npmmirror.com/mirrors/edgedriver",
    "ELECTRON_BUILDER_BINARIES_MIRROR": "https://cdn.npmmirror.com/binaries/electron-builder-binaries/",
    "ELECTRON_MIRROR": "https://cdn.npmmirror.com/binaries/electron/",
    "NODEJS_ORG_MIRROR": "https://cdn.npmmirror.com/binaries/node",
    "NVM_NODEJS_ORG_MIRROR": "https://cdn.npmmirror.com/binaries/node",
    "NWJS_URLBASE": "https://cdn.npmmirror.com/binaries/nwjs/v",
    "OPERADRIVER_CDNURL": "https://cdn.npmmirror.com/binaries/operadriver",
    "PHANTOMJS_CDNURL": "https://cdn.npmmirror.com/binaries/phantomjs",
    "PLAYWRIGHT_DOWNLOAD_HOST": "https://cdn.npmmirror.com/binaries/playwright",
    "PRISMA_ENGINES_MIRROR": "https://cdn.npmmirror.com/binaries/prisma",
    "PUPPETEER_DOWNLOAD_BASE_URL": "https://cdn.npmmirror.com/binaries/chrome-for-testing",
    "PUPPETEER_DOWNLOAD_HOST": "https://cdn.npmmirror.com/binaries/chrome-for-testing",
    "RE2_DOWNLOAD_MIRROR": "https://cdn.npmmirror.com/binaries/node-re2",
    "RE2_DOWNLOAD_SKIP_PATH": "true",
    "SASS_BINARY_SITE": "https://cdn.npmmirror.com/binaries/node-sass",
    "SAUCECTL_INSTALL_BINARY_MIRROR": "https://cdn.npmmirror.com/binaries/saucectl",
    "SENTRYCLI_CDNURL": "https://cdn.npmmirror.com/binaries/sentry-cli",
    "SWC_BINARY_SITE": "https://cdn.npmmirror.com/binaries/node-swc",
    "npm_config_better_sqlite3_binary_host": "https://cdn.npmmirror.com/binaries/better-sqlite3",
    "npm_config_keytar_binary_host": "https://cdn.npmmirror.com/binaries/keytar",
    "npm_config_robotjs_binary_host": "https://cdn.npmmirror.com/binaries/robotjs",
    "npm_config_sharp_binary_host": "https://cdn.npmmirror.com/binaries/sharp",
    "npm_config_sharp_libvips_binary_host": "https://cdn.npmmirror.com/binaries/sharp-libvips",
  },
  "terminal.integrated.env.osx": {
    "CHROMEDRIVER_CDNURL": "https://cdn.npmmirror.com/binaries/chromedriver",
    "COREPACK_NPM_REGISTRY": "https://registry.npmmirror.com",
    "CYPRESS_DOWNLOAD_PATH_TEMPLATE": "https://cdn.npmmirror.com/binaries/cypress/\${version}/\${platform}-\${arch}/cypress.zip",
    "EDGEDRIVER_CDNURL": "https://npmmirror.com/mirrors/edgedriver",
    "ELECTRON_BUILDER_BINARIES_MIRROR": "https://cdn.npmmirror.com/binaries/electron-builder-binaries/",
    "ELECTRON_MIRROR": "https://cdn.npmmirror.com/binaries/electron/",
    "NODEJS_ORG_MIRROR": "https://cdn.npmmirror.com/binaries/node",
    "NVM_NODEJS_ORG_MIRROR": "https://cdn.npmmirror.com/binaries/node",
    "NWJS_URLBASE": "https://cdn.npmmirror.com/binaries/nwjs/v",
    "OPERADRIVER_CDNURL": "https://cdn.npmmirror.com/binaries/operadriver",
    "PHANTOMJS_CDNURL": "https://cdn.npmmirror.com/binaries/phantomjs",
    "PLAYWRIGHT_DOWNLOAD_HOST": "https://cdn.npmmirror.com/binaries/playwright",
    "PRISMA_ENGINES_MIRROR": "https://cdn.npmmirror.com/binaries/prisma",
    "PUPPETEER_DOWNLOAD_BASE_URL": "https://cdn.npmmirror.com/binaries/chrome-for-testing",
    "PUPPETEER_DOWNLOAD_HOST": "https://cdn.npmmirror.com/binaries/chrome-for-testing",
    "RE2_DOWNLOAD_MIRROR": "https://cdn.npmmirror.com/binaries/node-re2",
    "RE2_DOWNLOAD_SKIP_PATH": "true",
    "SASS_BINARY_SITE": "https://cdn.npmmirror.com/binaries/node-sass",
    "SAUCECTL_INSTALL_BINARY_MIRROR": "https://cdn.npmmirror.com/binaries/saucectl",
    "SENTRYCLI_CDNURL": "https://cdn.npmmirror.com/binaries/sentry-cli",
    "SWC_BINARY_SITE": "https://cdn.npmmirror.com/binaries/node-swc",
    "npm_config_better_sqlite3_binary_host": "https://cdn.npmmirror.com/binaries/better-sqlite3",
    "npm_config_keytar_binary_host": "https://cdn.npmmirror.com/binaries/keytar",
    "npm_config_robotjs_binary_host": "https://cdn.npmmirror.com/binaries/robotjs",
    "npm_config_sharp_binary_host": "https://cdn.npmmirror.com/binaries/sharp",
    "npm_config_sharp_libvips_binary_host": "https://cdn.npmmirror.com/binaries/sharp-libvips",
  },
  "terminal.integrated.env.windows": {
    "CHROMEDRIVER_CDNURL": "https://cdn.npmmirror.com/binaries/chromedriver",
    "COREPACK_NPM_REGISTRY": "https://registry.npmmirror.com",
    "CYPRESS_DOWNLOAD_PATH_TEMPLATE": "https://cdn.npmmirror.com/binaries/cypress/\${version}/\${platform}-\${arch}/cypress.zip",
    "EDGEDRIVER_CDNURL": "https://npmmirror.com/mirrors/edgedriver",
    "ELECTRON_BUILDER_BINARIES_MIRROR": "https://cdn.npmmirror.com/binaries/electron-builder-binaries/",
    "ELECTRON_MIRROR": "https://cdn.npmmirror.com/binaries/electron/",
    "NODEJS_ORG_MIRROR": "https://cdn.npmmirror.com/binaries/node",
    "NVM_NODEJS_ORG_MIRROR": "https://cdn.npmmirror.com/binaries/node",
    "NWJS_URLBASE": "https://cdn.npmmirror.com/binaries/nwjs/v",
    "OPERADRIVER_CDNURL": "https://cdn.npmmirror.com/binaries/operadriver",
    "PHANTOMJS_CDNURL": "https://cdn.npmmirror.com/binaries/phantomjs",
    "PLAYWRIGHT_DOWNLOAD_HOST": "https://cdn.npmmirror.com/binaries/playwright",
    "PRISMA_ENGINES_MIRROR": "https://cdn.npmmirror.com/binaries/prisma",
    "PUPPETEER_DOWNLOAD_BASE_URL": "https://cdn.npmmirror.com/binaries/chrome-for-testing",
    "PUPPETEER_DOWNLOAD_HOST": "https://cdn.npmmirror.com/binaries/chrome-for-testing",
    "RE2_DOWNLOAD_MIRROR": "https://cdn.npmmirror.com/binaries/node-re2",
    "RE2_DOWNLOAD_SKIP_PATH": "true",
    "SASS_BINARY_SITE": "https://cdn.npmmirror.com/binaries/node-sass",
    "SAUCECTL_INSTALL_BINARY_MIRROR": "https://cdn.npmmirror.com/binaries/saucectl",
    "SENTRYCLI_CDNURL": "https://cdn.npmmirror.com/binaries/sentry-cli",
    "SWC_BINARY_SITE": "https://cdn.npmmirror.com/binaries/node-swc",
    "npm_config_better_sqlite3_binary_host": "https://cdn.npmmirror.com/binaries/better-sqlite3",
    "npm_config_keytar_binary_host": "https://cdn.npmmirror.com/binaries/keytar",
    "npm_config_robotjs_binary_host": "https://cdn.npmmirror.com/binaries/robotjs",
    "npm_config_sharp_binary_host": "https://cdn.npmmirror.com/binaries/sharp",
    "npm_config_sharp_libvips_binary_host": "https://cdn.npmmirror.com/binaries/sharp-libvips",
  },
}
`;
